File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.swift
import Foundation

let n = Int(readLine()!)!
var triangle = [[Int]]()
for _ in 0..<n {
    triangle.append(readLine()!.split(separator: " ").map { Int($0)! })
}
var dp = Array(repeating: Array(repeating: 0, count: n), count: n)
dp[0][0] = triangle[0][0]

for i in 1..<n {
    dp[i][0] = dp[i-1][0] + triangle[i][0]
    if n > 2 {
        for j in 1..<i {
            dp[i][j] = max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j]
        }
    }
    dp[i][i] = dp[i-1][i-1] + triangle[i][i]
}

print(dp[n-1].max()!)

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    length = len(triangle)
    dp = [[0] * length for _ in range(length)]
    dp[0][0] = triangle[0][0]

    for i in range(1, length):
        for j in range(i+1):
            dp[i][j] = triangle[i][j]
            if j == 0:
                dp[i][j] += dp[i-1][j]
            elif j == i:
                dp[i][j] += dp[i-1][j-1]
            else:
                dp[i][j] += max(dp[i-1][j], dp[i-1][j-1])

    return max(dp[length - 1])